### PR TITLE
Feature: Delegation of the RIB MA-IPCP

### PR DESCRIPTION
**Pull request by bernatgaston**
A new feature that allows the MA to receive a CDAP message for a
managed IPCP. THe MA can send this message to the corresponding IPCP.
The response is catch again by the MA who can send it to the
manager.

ACK: @lbergesio @edugrasa @vmaffione @vicalro

### CODE REVIEW

**Comment by vmaffione** [comment]
I need to take a look in depth, but is this based on the work I started some time ago (netlink support for RIB delegation, #617 #624) ?

**Comment by bernatgaston** [comment]
Yes, basically it now links the MA with the IPCP using the mentioned work

**Comment by vmaffione** [comment]
installing from scratch, rinad build fails

````
make[4]: Entering directory '/home/vmaffione/irati/build/rinad/src/common/encoders'
rm -f ApplicationProcessNamingInfoMessage.tmp
touch ApplicationProcessNamingInfoMessage.tmp
/usr/bin/protoc -I. -I/home/vmaffione/irati/rinad/src/common/encoders	\
	--cpp_out=.		\
	/home/vmaffione/irati/rinad/src/common/encoders/ApplicationProcessNamingInfoMessage.proto
mv -f ApplicationProcessNamingInfoMessage.tmp ApplicationProcessNamingInfoMessage.stamp
if test -f ApplicationProcessNamingInfoMessage.pb.cc; then :; else \
  trap 'rm -rf ApplicationProcessNamingInfoMessage.lock ApplicationProcessNamingInfoMessage.stamp' 1 2 13 15; \
  if mkdir ApplicationProcessNamingInfoMessage.lock 2>/dev/null; then \
    rm -f ApplicationProcessNamingInfoMessage.stamp; \
    make  ApplicationProcessNamingInfoMessage.stamp; \
    result=$?; rm -rf ApplicationProcessNamingInfoMessage.lock; exit $result; \
  else \
    while test -d ApplicationProcessNamingInfoMessage.lock; do sleep 1; done; \
    test -f ApplicationProcessNamingInfoMessage.stamp; \
  fi; \
fi
rm -f CommonMessages.tmp
touch CommonMessages.tmp
/usr/bin/protoc -I. -I/home/vmaffione/irati/rinad/src/common/encoders	\
	--cpp_out=.		\
	/home/vmaffione/irati/rinad/src/common/encoders/CommonMessages.proto
mv -f CommonMessages.tmp CommonMessages.stamp
if test -f CommonMessages.pb.cc; then :; else \
  trap 'rm -rf CommonMessages.lock CommonMessages.stamp' 1 2 13 15; \
  if mkdir CommonMessages.lock 2>/dev/null; then \
    rm -f CommonMessages.stamp; \
    make  CommonMessages.stamp; \
    result=$?; rm -rf CommonMessages.lock; exit $result; \
  else \
    while test -d CommonMessages.lock; do sleep 1; done; \
    test -f CommonMessages.stamp; \
  fi; \
fi
rm -f ConnectionPoliciesMessage.tmp
touch ConnectionPoliciesMessage.tmp
/usr/bin/protoc -I. -I/home/vmaffione/irati/rinad/src/common/encoders	\
	--cpp_out=.		\
	/home/vmaffione/irati/rinad/src/common/encoders/ConnectionPoliciesMessage.proto
mv -f ConnectionPoliciesMessage.tmp ConnectionPoliciesMessage.stamp
if test -f ConnectionPoliciesMessage.pb.cc; then :; else \
  trap 'rm -rf ConnectionPoliciesMessage.lock ConnectionPoliciesMessage.stamp' 1 2 13 15; \
  if mkdir ConnectionPoliciesMessage.lock 2>/dev/null; then \
    rm -f ConnectionPoliciesMessage.stamp; \
    make  ConnectionPoliciesMessage.stamp; \
    result=$?; rm -rf ConnectionPoliciesMessage.lock; exit $result; \
  else \
    while test -d ConnectionPoliciesMessage.lock; do sleep 1; done; \
    test -f ConnectionPoliciesMessage.stamp; \
  fi; \
fi
rm -f DataTransferConstantsMessage.tmp
touch DataTransferConstantsMessage.tmp
/usr/bin/protoc -I. -I/home/vmaffione/irati/rinad/src/common/encoders	\
	--cpp_out=.		\
	/home/vmaffione/irati/rinad/src/common/encoders/DataTransferConstantsMessage.proto
mv -f DataTransferConstantsMessage.tmp DataTransferConstantsMessage.stamp
if test -f DataTransferConstantsMessage.pb.cc; then :; else \
  trap 'rm -rf DataTransferConstantsMessage.lock DataTransferConstantsMessage.stamp' 1 2 13 15; \
  if mkdir DataTransferConstantsMessage.lock 2>/dev/null; then \
    rm -f DataTransferConstantsMessage.stamp; \
    make  DataTransferConstantsMessage.stamp; \
    result=$?; rm -rf DataTransferConstantsMessage.lock; exit $result; \
  else \
    while test -d DataTransferConstantsMessage.lock; do sleep 1; done; \
    test -f DataTransferConstantsMessage.stamp; \
  fi; \
fi
rm -f DirectoryForwardingTableEntryArrayMessage.tmp
touch DirectoryForwardingTableEntryArrayMessage.tmp
/usr/bin/protoc -I. -I/home/vmaffione/irati/rinad/src/common/encoders	\
	--cpp_out=.		\
	/home/vmaffione/irati/rinad/src/common/encoders/DirectoryForwardingTableEntryArrayMessage.proto
mv -f DirectoryForwardingTableEntryArrayMessage.tmp DirectoryForwardingTableEntryArrayMessage.stamp
if test -f DirectoryForwardingTableEntryArrayMessage.pb.cc; then :; else \
  trap 'rm -rf DirectoryForwardingTableEntryArrayMessage.lock DirectoryForwardingTableEntryArrayMessage.stamp' 1 2 13 15; \
  if mkdir DirectoryForwardingTableEntryArrayMessage.lock 2>/dev/null; then \
    rm -f DirectoryForwardingTableEntryArrayMessage.stamp; \
    make  DirectoryForwardingTableEntryArrayMessage.stamp; \
    result=$?; rm -rf DirectoryForwardingTableEntryArrayMessage.lock; exit $result; \
  else \
    while test -d DirectoryForwardingTableEntryArrayMessage.lock; do sleep 1; done; \
    test -f DirectoryForwardingTableEntryArrayMessage.stamp; \
  fi; \
fi
rm -f DirectoryForwardingTableEntryMessage.tmp
touch DirectoryForwardingTableEntryMessage.tmp
/usr/bin/protoc -I. -I/home/vmaffione/irati/rinad/src/common/encoders	\
	--cpp_out=.		\
	/home/vmaffione/irati/rinad/src/common/encoders/DirectoryForwardingTableEntryMessage.proto
mv -f DirectoryForwardingTableEntryMessage.tmp DirectoryForwardingTableEntryMessage.stamp
if test -f DirectoryForwardingTableEntryMessage.pb.cc; then :; else \
  trap 'rm -rf DirectoryForwardingTableEntryMessage.lock DirectoryForwardingTableEntryMessage.stamp' 1 2 13 15; \
  if mkdir DirectoryForwardingTableEntryMessage.lock 2>/dev/null; then \
    rm -f DirectoryForwardingTableEntryMessage.stamp; \
    make  DirectoryForwardingTableEntryMessage.stamp; \
    result=$?; rm -rf DirectoryForwardingTableEntryMessage.lock; exit $result; \
  else \
    while test -d DirectoryForwardingTableEntryMessage.lock; do sleep 1; done; \
    test -f DirectoryForwardingTableEntryMessage.stamp; \
  fi; \
fi
rm -f EnrollmentInformationMessage.tmp
touch EnrollmentInformationMessage.tmp
/usr/bin/protoc -I. -I/home/vmaffione/irati/rinad/src/common/encoders	\
	--cpp_out=.		\
	/home/vmaffione/irati/rinad/src/common/encoders/EnrollmentInformationMessage.proto
mv -f EnrollmentInformationMessage.tmp EnrollmentInformationMessage.stamp
if test -f EnrollmentInformationMessage.pb.cc; then :; else \
  trap 'rm -rf EnrollmentInformationMessage.lock EnrollmentInformationMessage.stamp' 1 2 13 15; \
  if mkdir EnrollmentInformationMessage.lock 2>/dev/null; then \
    rm -f EnrollmentInformationMessage.stamp; \
    make  EnrollmentInformationMessage.stamp; \
    result=$?; rm -rf EnrollmentInformationMessage.lock; exit $result; \
  else \
    while test -d EnrollmentInformationMessage.lock; do sleep 1; done; \
    test -f EnrollmentInformationMessage.stamp; \
  fi; \
fi
rm -f FlowMessage.tmp
touch FlowMessage.tmp
/usr/bin/protoc -I. -I/home/vmaffione/irati/rinad/src/common/encoders	\
	--cpp_out=.		\
	/home/vmaffione/irati/rinad/src/common/encoders/FlowMessage.proto
mv -f FlowMessage.tmp FlowMessage.stamp
if test -f FlowMessage.pb.cc; then :; else \
  trap 'rm -rf FlowMessage.lock FlowMessage.stamp' 1 2 13 15; \
  if mkdir FlowMessage.lock 2>/dev/null; then \
    rm -f FlowMessage.stamp; \
    make  FlowMessage.stamp; \
    result=$?; rm -rf FlowMessage.lock; exit $result; \
  else \
    while test -d FlowMessage.lock; do sleep 1; done; \
    test -f FlowMessage.stamp; \
  fi; \
fi
rm -f FlowStateGroupMessage.tmp
touch FlowStateGroupMessage.tmp
/usr/bin/protoc -I. -I/home/vmaffione/irati/rinad/src/common/encoders	\
	--cpp_out=.		\
	/home/vmaffione/irati/rinad/src/common/encoders/FlowStateGroupMessage.proto
mv -f FlowStateGroupMessage.tmp FlowStateGroupMessage.stamp
if test -f FlowStateGroupMessage.pb.cc; then :; else \
  trap 'rm -rf FlowStateGroupMessage.lock FlowStateGroupMessage.stamp' 1 2 13 15; \
  if mkdir FlowStateGroupMessage.lock 2>/dev/null; then \
    rm -f FlowStateGroupMessage.stamp; \
    make  FlowStateGroupMessage.stamp; \
    result=$?; rm -rf FlowStateGroupMessage.lock; exit $result; \
  else \
    while test -d FlowStateGroupMessage.lock; do sleep 1; done; \
    test -f FlowStateGroupMessage.stamp; \
  fi; \
fi
rm -f FlowStateMessage.tmp
touch FlowStateMessage.tmp
/usr/bin/protoc -I. -I/home/vmaffione/irati/rinad/src/common/encoders	\
	--cpp_out=.		\
	/home/vmaffione/irati/rinad/src/common/encoders/FlowStateMessage.proto
mv -f FlowStateMessage.tmp FlowStateMessage.stamp
if test -f FlowStateMessage.pb.cc; then :; else \
  trap 'rm -rf FlowStateMessage.lock FlowStateMessage.stamp' 1 2 13 15; \
  if mkdir FlowStateMessage.lock 2>/dev/null; then \
    rm -f FlowStateMessage.stamp; \
    make  FlowStateMessage.stamp; \
    result=$?; rm -rf FlowStateMessage.lock; exit $result; \
  else \
    while test -d FlowStateMessage.lock; do sleep 1; done; \
    test -f FlowStateMessage.stamp; \
  fi; \
fi
rm -f MA-IPCP.tmp
touch MA-IPCP.tmp
/usr/bin/protoc -I. -I/home/vmaffione/irati/rinad/src/common/encoders	\
	--cpp_out=.		\
	/home/vmaffione/irati/rinad/src/common/encoders/MA-IPCP.proto
mv -f MA-IPCP.tmp MA-IPCP.stamp
if test -f MA-IPCP.pb.cc; then :; else \
  trap 'rm -rf MA-IPCP.lock MA-IPCP.stamp' 1 2 13 15; \
  if mkdir MA-IPCP.lock 2>/dev/null; then \
    rm -f MA-IPCP.stamp; \
    make  MA-IPCP.stamp; \
    result=$?; rm -rf MA-IPCP.lock; exit $result; \
  else \
    while test -d MA-IPCP.lock; do sleep 1; done; \
    test -f MA-IPCP.stamp; \
  fi; \
fi
rm -f NeighborArrayMessage.tmp
touch NeighborArrayMessage.tmp
/usr/bin/protoc -I. -I/home/vmaffione/irati/rinad/src/common/encoders	\
	--cpp_out=.		\
	/home/vmaffione/irati/rinad/src/common/encoders/NeighborArrayMessage.proto
mv -f NeighborArrayMessage.tmp NeighborArrayMessage.stamp
if test -f NeighborArrayMessage.pb.cc; then :; else \
  trap 'rm -rf NeighborArrayMessage.lock NeighborArrayMessage.stamp' 1 2 13 15; \
  if mkdir NeighborArrayMessage.lock 2>/dev/null; then \
    rm -f NeighborArrayMessage.stamp; \
    make  NeighborArrayMessage.stamp; \
    result=$?; rm -rf NeighborArrayMessage.lock; exit $result; \
  else \
    while test -d NeighborArrayMessage.lock; do sleep 1; done; \
    test -f NeighborArrayMessage.stamp; \
  fi; \
fi
rm -f NeighborMessage.tmp
touch NeighborMessage.tmp
/usr/bin/protoc -I. -I/home/vmaffione/irati/rinad/src/common/encoders	\
	--cpp_out=.		\
	/home/vmaffione/irati/rinad/src/common/encoders/NeighborMessage.proto
mv -f NeighborMessage.tmp NeighborMessage.stamp
if test -f NeighborMessage.pb.cc; then :; else \
  trap 'rm -rf NeighborMessage.lock NeighborMessage.stamp' 1 2 13 15; \
  if mkdir NeighborMessage.lock 2>/dev/null; then \
    rm -f NeighborMessage.stamp; \
    make  NeighborMessage.stamp; \
    result=$?; rm -rf NeighborMessage.lock; exit $result; \
  else \
    while test -d NeighborMessage.lock; do sleep 1; done; \
    test -f NeighborMessage.stamp; \
  fi; \
fi
rm -f PolicyDescriptorMessage.tmp
touch PolicyDescriptorMessage.tmp
/usr/bin/protoc -I. -I/home/vmaffione/irati/rinad/src/common/encoders	\
	--cpp_out=.		\
	/home/vmaffione/irati/rinad/src/common/encoders/PolicyDescriptorMessage.proto
mv -f PolicyDescriptorMessage.tmp PolicyDescriptorMessage.stamp
if test -f PolicyDescriptorMessage.pb.cc; then :; else \
  trap 'rm -rf PolicyDescriptorMessage.lock PolicyDescriptorMessage.stamp' 1 2 13 15; \
  if mkdir PolicyDescriptorMessage.lock 2>/dev/null; then \
    rm -f PolicyDescriptorMessage.stamp; \
    make  PolicyDescriptorMessage.stamp; \
    result=$?; rm -rf PolicyDescriptorMessage.lock; exit $result; \
  else \
    while test -d PolicyDescriptorMessage.lock; do sleep 1; done; \
    test -f PolicyDescriptorMessage.stamp; \
  fi; \
fi
rm -f QoSCubeArrayMessage.tmp
touch QoSCubeArrayMessage.tmp
/usr/bin/protoc -I. -I/home/vmaffione/irati/rinad/src/common/encoders	\
	--cpp_out=.		\
	/home/vmaffione/irati/rinad/src/common/encoders/QoSCubeArrayMessage.proto
mv -f QoSCubeArrayMessage.tmp QoSCubeArrayMessage.stamp
if test -f QoSCubeArrayMessage.pb.cc; then :; else \
  trap 'rm -rf QoSCubeArrayMessage.lock QoSCubeArrayMessage.stamp' 1 2 13 15; \
  if mkdir QoSCubeArrayMessage.lock 2>/dev/null; then \
    rm -f QoSCubeArrayMessage.stamp; \
    make  QoSCubeArrayMessage.stamp; \
    result=$?; rm -rf QoSCubeArrayMessage.lock; exit $result; \
  else \
    while test -d QoSCubeArrayMessage.lock; do sleep 1; done; \
    test -f QoSCubeArrayMessage.stamp; \
  fi; \
fi
rm -f QoSCubeMessage.tmp
touch QoSCubeMessage.tmp
/usr/bin/protoc -I. -I/home/vmaffione/irati/rinad/src/common/encoders	\
	--cpp_out=.		\
	/home/vmaffione/irati/rinad/src/common/encoders/QoSCubeMessage.proto
mv -f QoSCubeMessage.tmp QoSCubeMessage.stamp
if test -f QoSCubeMessage.pb.cc; then :; else \
  trap 'rm -rf QoSCubeMessage.lock QoSCubeMessage.stamp' 1 2 13 15; \
  if mkdir QoSCubeMessage.lock 2>/dev/null; then \
    rm -f QoSCubeMessage.stamp; \
    make  QoSCubeMessage.stamp; \
    result=$?; rm -rf QoSCubeMessage.lock; exit $result; \
  else \
    while test -d QoSCubeMessage.lock; do sleep 1; done; \
    test -f QoSCubeMessage.stamp; \
  fi; \
fi
rm -f QoSSpecification.tmp
touch QoSSpecification.tmp
/usr/bin/protoc -I. -I/home/vmaffione/irati/rinad/src/common/encoders	\
	--cpp_out=.		\
	/home/vmaffione/irati/rinad/src/common/encoders/QoSSpecification.proto
mv -f QoSSpecification.tmp QoSSpecification.stamp
if test -f QoSSpecification.pb.cc; then :; else \
  trap 'rm -rf QoSSpecification.lock QoSSpecification.stamp' 1 2 13 15; \
  if mkdir QoSSpecification.lock 2>/dev/null; then \
    rm -f QoSSpecification.stamp; \
    make  QoSSpecification.stamp; \
    result=$?; rm -rf QoSSpecification.lock; exit $result; \
  else \
    while test -d QoSSpecification.lock; do sleep 1; done; \
    test -f QoSSpecification.stamp; \
  fi; \
fi
make[4]: *** No rule to make target 'RIBObjectData.proto', needed by 'RIBObjectData.stamp'.  Stop.
make[4]: Leaving directory '/home/vmaffione/irati/build/rinad/src/common/encoders'
Makefile:701: recipe for target 'install-recursive' failed
make[3]: *** [install-recursive] Error 1
make[3]: Leaving directory '/home/vmaffione/irati/build/rinad/src/common'
Makefile:406: recipe for target 'install-recursive' failed
make[2]: *** [install-recursive] Error 1
make[2]: Leaving directory '/home/vmaffione/irati/build/rinad/src'
Makefile:441: recipe for target 'install-recursive' failed
make[1]: *** [install-recursive] Error 1
make[1]: Leaving directory '/home/vmaffione/irati/build/rinad'
Makefile:738: recipe for target 'install' failed
make: *** [install] Error 2
install-user-from-scratch: Cannot complete rinad phase
````

File: librina/include/librina/ipc-daemons.h [hunk 1]
```diff
@@ -198,10 +198,28 @@ class FwdCDAPMsgEvent: public IPCEvent {
...
+	FwdCDAPMsgRequestEvent(const ser_obj_t& sm, int result,
 			unsigned int sequenceNumber);
 };
 
+/**
+ * The IPC Manager wants to forward a CDAP response message to
+ * an IPC process, or the IPC process wants to forwards a
+ * CDAP response back to the IPC Manager.
+ */
+class FwdCDAPMsgResponseEvent: public IPCEvent {
```
Comment (vmaffione) [inline librina/include/librina/ipc-daemons.h]: This class FwdCDAPMsgResponseEvent is just a clone of FwdCDAPMsgRequestEvent. Do we need it?

**Comment by bernatgaston** [comment]
Sorry I was missing a file. Now it is added. I forced the push

**Comment by vmaffione** [comment]
Ok, now it compiles. My regression tests are ok. But how can I see the delegation in action?

**Comment by bernatgaston** [comment]
that's more difficult. You need a manager to test it. You can do it by using the general manager (which does not have documentation for cdap) or you can try to use the minimanager https://github.com/bernatgaston/manager
But you will also need the configurations so it is a bit difficult if you are not used to it. If you really need to test it I will try to send specific instructions, but I don't want to pospone the PR too much

**Comment by edugrasa** [comment]
RIBObjectDataList is not included in the file https://github.com/bernatgaston/stack/blob/ipcp_delegation/rinad/src/common/encoders/Makefile.inc#L596, as a consequence the GPB message is not compiled and no C++ stub classes generated (then the compilation of the mini manager fails). Could you fix it?

**Comment by vmaffione** [comment]
@bernatgaston no worries I don't want to delay this.I've already tried out the minimanager in the past. Is that compiled automatically in rinad?

**Comment by bernatgaston** [comment]
Minimanager fixed, please @edugrasa try again

**Comment by edugrasa** [comment]
Testing done, works well! Merging the PR, since Vincenzo verified there were no regressions.

**Pull request merged**